pytential/symbolic/primitives.py

```diff
@@ -966,7 +988,29 @@ def _simplex_mapping_max_stretch_factor(ambient_dim, dim=None, dofdesc=None,
     if with_elementwise_max:
         result = ElementwiseMax(result, dofdesc=dofdesc)
 
-    return cse(result, "mapping_max_stretch", cse_scope.DISCRETIZATION)
+    return cse(result, "simplex_mapping_max_stretch", cse_scope.DISCRETIZATION)
+
+
+def _hypercube_mapping_max_stretch_factor(ambient_dim, dim=None, dofdesc=None,
+        with_elementwise_max=True):
+    if dim is None:
+        dim = ambient_dim - 1
+
+    # NOTE: unlike in the simplex case, here we do not need to transform the
+    # reference element, as it already is nicely rotation invariant
+    pder_mat = first_fundamental_form(ambient_dim, dim=dim, dofdesc=dofdesc)
+    stretch_factors = [
+            cse(4.0 * sqrt(s), f"hypercube_mapping_singval_{i}")
```

Comment:
Besides the obvious hackery of `_ShapeDiscretizationProperty`, I'm not quite sure where this `4` comes from. Without adding a factor here, it wouldn't refine anything on the spheroid case. I imagine it has something to do with simplices vs quads in the actual magnitude (2 simplices make up a quad?), but no clue really.

Any ideas?

Reply:
The 4 in the simplex case corrects for the fact that the reference element is [-1,1], which shows up twice in J^T J (in case that wasn't clear from the comment). It would also show up twice in the first fundamental form.

What this function should do is faithfully approximate the max "stretch factor" vs a unit-size reference element. Note that the simplex case has the four *inside* the square root, the one here is outside. I think a 2 outside or a 4 inside is justifiable.

Reply:
(And if, based on that, it decides no stage-2 refinement is needed, then great.)

Reply:
Went ahead and changed that to a factor of 4 inside, exactly like for the simplex case.